### PR TITLE
fix(attention): stop handing FlashAttention 4 the -1 window sentinel

### LINKED ISSUE #3528: FA4 returns an all-zero output for causal attention on SM100 after flash-attention #2490
## Summary

With a flash-attn-4 build that includes [#2490](https://github.com/Dao-AILab/flash-attention/pull/2490),
FlashAttention 4 returns an **all-zero output** for causal attention on SM100 when called the way
TransformerEngine calls it. No context parallelism is needed to reproduce it. It is silent: no
exception, no warning, finite values, every element zero.

Context parallelism is only where it was first noticed, because the p2p ring merges the
accompanying `-inf` log-sum-exp and turns a silent wrong answer into a visible NaN.

> **Edited 2026-09-17.** This issue originally described a CP-specific NaN and, in a later edit,
> argued the window encoding was not at fault. Both were wrong. Direct measurement against a float64
> reference shows the defect is general to FA4 causal attention on SM100. History of both earlier
> readings is kept at the bottom. Fixed by #3532, which also repairs head_dim 512 under context
> parallelism -- see below; the earlier suggestion that 512 might be a second defect is retracted.

## Reproduction, without context parallelism

```python
import torch, transformer_engine.pytorch as te
b, s, h, d = 2, 1024, 8, 128
q, k, v = (torch.randn(b, s, h, d, device="cuda", dtype=torch.bfloat16) for _ in range(3))
dpa = te.DotProductAttention(h, d, qkv_format="bshd", attn_mask_type="causal",
                             attention_dropout=0.0).to(dtype=torch.bfloat16, device="cuda")
print(dpa(q, k, v).abs().max())   # tensor(0., device='cuda:0')
```

Run with `NVTE_FUSED_ATTN=0 NVTE_UNFUSED_ATTN=0` so the selection cannot fall through to cuDNN.
Measured on B200:

```
Selected backend = FlashAttention (4.0.0b31.dev9+g346aa9e)
context_parallel: False, cp_size: 1
  flash_attn_func_v4  causal=True  window_size=(-1, 0)
  out.abs().max()=0.000000e+00  zeros=YES  relerr vs float64 reference = 1.000e+00
```

With `NVTE_FLASH_ATTN_V4=0` the same script selects FlashAttention 2 and returns
`out.abs().max()=3.453125e+00`, `relerr=2.046e-03`. The reference is float64, not float32: torch
uses TF32 for fp32 matmuls on Ampere and newer, and TF32 carries an 11-bit significand, the same as
fp16, so an fp32 reference cannot judge a bf16 kernel.

## The kernel, with TransformerEngine out of the picture

```python
from flash_attn.cute.interface import flash_attn_func
flash_attn_func(q, k, v, causal=True, window_size=(-1, 0))     # out.abs().max() = 0.0
flash_attn_func(q, k, v, causal=True, window_size=(None, None))# out.abs().max() = 3.453125, relerr 2.0e-03
```

## Why

TransformerEngine normalises causal masking to `window_size = (-1, 0)`
(`dot_product_attention/utils.py`, `check_set_window_size`) and passes it to FA4, whose sentinel for
"unbounded" is `None`, not `-1`. Dao-AILab/flash-attention#2490 ("cute: don't widen intentionally-empty offset windows to full
attention", merged 2026-09-09) narrowed the shim that used to absorb the mismatch:

```python
# before #2490 -- causal applied first, then widen if the SUM is negative
if causal: window_size_right = 0
if wsl is not None and wsr is not None and wsl + wsr < 0: wsl = wsr = None

# after #2490 -- widen only if BOTH are negative, then apply causal
if wsl is not None and wsr is not None and wsl < 0 and wsr < 0: wsl = wsr = None
if causal: window_size_right = 0
```

Under the old rule `(-1, 0)` summed to `-1` and always collapsed to `(None, None)`, so TE's window
encoding was discarded and `causal=True` alone set the mask. Under the new rule `(-1, 0)` survives:
`_resolve_causal_local_window(True, -1, 0)` returns `(False, True, -1, 0)`, so causal is dropped and a
local band is applied with left bound `-1`. `mask.py` guards on `window_size_left is not None` and
uses the value arithmetically, giving a band of `[row+1, row]` -- empty. Measured directly:

```
lse   n=131072  -inf=131072  +inf=0  nan=0     every element
out   n=16777216  -inf=0  +inf=0  nan=0        every element zero
```

FA4 is internally inconsistent about this: the d512 dispatch at `interface.py:783` tests
`window_size_left in (None, -1)`, treating `-1` as "no window", while the resolver treats it as a
bound. But TE should not be sending `-1` regardless.

## Why CI and the CP tests do not catch it

Three independent reasons, all worth fixing on their own:

1. `NVTE_FLASH_ATTN_V4=0` is exported in `qa/L1_pytorch_distributed_unittest/test.sh` and the L0
   suites, so FA4 is never exercised under context parallelism.
2. CI pins `flash-attn-4==4.0.0b11`, which predates Dao-AILab/flash-attention#2490.
3. **`run_attention_with_cp.py` compares a CP run against a non-CP run of the same backend**
   (`tensors_no_cp` vs `tensors_cp`). When both sides return zeros they agree, and the test passes
   while measuring nothing. This is why `all_gather` and `a2a` report PASS on a backend that is
   returning zeros -- a systematic error appears identically on both sides and cancels. Only `p2p`
   fails, because the ring merge computes `exp(lse_step - lse)` over `-inf` and produces NaN, which
   then differs from the zeros on the other side.

Point 3 means the CP suite cannot detect any error that affects CP and non-CP equally. That is a
gap independent of this bug.

## Scope

Any configuration on SM100 where TE selects FA4 for a causal mask, with a flash-attn-4 that includes
Dao-AILab/flash-attention#2490. cuDNN FusedAttention usually wins backend selection for common shapes, which is likely why
this has gone unnoticed -- but it does not win where cuDNN has no engine, which includes large head
dims, and it does not win when fused attention is disabled.

## Suggested fix

TE should express "unbounded" to FA4 as `None` rather than `-1`, at **every** FA4 entry point --
both the public `flash_attn_func_v4` used on the non-CP path and `_flash_attn_fwd_v4` /
`_flash_attn_bwd_v4` used by context parallelism. Genuine sliding windows already pass non-negative
bounds and are unaffected.

A partial fix is worse than none here, and this is not hypothetical: I first tried translating the
sentinel only at `_flash_attn_fwd_v4` / `_flash_attn_bwd_v4`, which corrected the CP side while
leaving the non-CP reference side returning zeros. The self-referential comparison above then
reported the *fix* as a regression in `all_gather` and `a2a`.

## head_dim 512 (previously flagged as possibly a second defect -- it is not)

At symmetric `head_dim=512`, with flash-attention Dao-AILab/flash-attention#2877 supplying the
kernels and TE's d512 gate lifted, all three `cp_comm_type` values failed with `dq has nan values`.
I originally suggested that might be an independent defect, reasoning that `all_gather` and `a2a`
perform no cross-step LSE merge so the ring correction could not be the source.

That reasoning was wrong. The **backward** consumes the log-sum-exp regardless of comm type, so an
all `-inf` LSE poisons `exp(scores - lse)` in any configuration. Measured as an A/B on one build,
one set of kernels, with the normalizer switched off and on at runtime:

| `cp_comm_type` | sentinel normalization off | on |
|---|---|---|
| `p2p` | FAIL `dq has nan values` | **PASS** |
| `all_gather` | FAIL `dq has nan values` | **PASS** |
| `a2a` | FAIL `dq has nan values` | **PASS** |

and at the kernel level, with no TransformerEngine involved:

```
head_dim 512, causal, window_size=(-1, 0)      out.abs().max() = 0.0
head_dim 512, causal, window_size=(None, None) out.abs().max() = 3.875
```

Same empty band as head_dim 128. One root cause, one fix. At head_dim 128 the degenerate input
yields zeros rather than NaN, which is why those arms passed against an equally-zero reference.

## Ruled out

- **TE passing uninitialised `dq`.** `context_parallel.py` uses `torch.empty_like` for the gradient
  buffers; patched to `zeros_like` and the NaN persisted. FA4's backward substitutes its own work
  buffer and copies back.
- **An LSE convention mismatch.** FA4's forward LSE measured at `(2, 4, 512)` against an fp32
  reference: natural log, max error 1.3e-5.

## Earlier readings, kept for the record

The first version of this issue attributed the NaN to the ring correction consuming an `-inf` LSE.
That mechanism is real and is why `p2p` fails loudly, but it is a consequence, not the root cause,
and it described the problem as CP-specific when it is not.

A later edit argued the window encoding was *not* at fault, reasoning that `all_gather` and `a2a`
send the same `(-1, 0)` and pass. That inference was wrong: they pass because the test compares them
against an equally broken reference, not because their output is correct.

## Environment

- B200 (SM100), `nvcr.io/nvidia/pytorch:26.08-py3`, cuDNN 9.25.0
- TransformerEngine 2.18.0
- flash-attn-4 `4.0.0b31.dev9+g346aa9e`; any build including Dao-AILab/flash-attention#2490 should reproduce the non-CP case

### PR DESCRIPTION
**Pull request by nvegesna-netizen**
## Problem

FlashAttention 4 returns an **all-zero output** for causal attention on SM100, silently — no
exception, no warning, finite values, every element zero. Fixes #3528.

No context parallelism is needed to reproduce it:

```python
import torch, transformer_engine.pytorch as te
b, s, h, d = 2, 1024, 8, 128
q, k, v = (torch.randn(b, s, h, d, device="cuda", dtype=torch.bfloat16) for _ in range(3))
dpa = te.DotProductAttention(h, d, qkv_format="bshd", attn_mask_type="causal",
                             attention_dropout=0.0).to(dtype=torch.bfloat16, device="cuda")
print(dpa(q, k, v).abs().max())   # tensor(0., device='cuda:0')
```

Run it with `NVTE_FUSED_ATTN=0 NVTE_UNFUSED_ATTN=0`, or cuDNN FusedAttention wins backend
selection for this shape and hides the defect entirely.

## Cause

TE normalises an unbounded window side to `-1` (`check_set_window_size`); FA4's sentinel is `None`.
Until Dao-AILab/flash-attention#2490 a shim widened any pair summing below zero to full attention,
so TE's causal encoding `(-1, 0)` was always erased and `causal=True` alone set the mask. #2490
narrowed that to pairs where **both** bounds are negative, so `(-1, 0)` now survives as a literal
band of `[row + 1, row]` — empty. `flash_attn/cute/mask.py` guards on `is not None` and uses the
value arithmetically, so the kernel returns zeros and an all `-inf` LSE.

Measured directly, with TransformerEngine out of the picture:

```python
flash_attn_func(q, k, v, causal=True, window_size=(-1, 0))      # out.abs().max() = 0.0
flash_attn_func(q, k, v, causal=True, window_size=(None, None)) # out.abs().max() = 3.453125
```

## Change

Normalise negative window bounds to `None` at the FA4 entry points, where the symbols are bound.

**Why there rather than at the call sites.** The ~20 sites that build these kwargs are shared with
FA2 >= 2.7 and FA3, for which `-1` is the correct spelling; rewriting them in place would mean
forking every one on `use_flash_attn_4`.

**Why all four entry points.** The non-CP path calls `flash_attn_func` / `flash_attn_varlen_func`;
context parallelism calls `_flash_attn_fwd` / `_flash_attn_bwd`. Covering only one pair is worse
than covering neither, because `run_attention_with_cp.py` grades a CP run against a non-CP run of
the same backend: with both sides returning zeros the comparison agrees and passes, so correcting
one side alone makes the fix present as a regression. That is not hypothetical — an earlier partial
attempt did exactly that, turning two passing arms red.

Genuine sliding windows already carry non-negative bounds and pass through untouched.

## Verification (B200, SM100)

| check | result |
|---|---|
| non-CP causal through `DotProductAttention` | `3.453125e+00`, was `0.000000e+00`; relerr vs float64 `2.046e-03` |
| sliding windows `(256,0)`, `(128,0)`, `(0,0)` | relerr `2.046e-03`, `2.046e-03`, `0.000e+00` |
| CP with FA4 on: `p2p` bshd/sbhd/thd | 3/3 pass — **`p2p` failed before this change** |
| CP with FA4 on: `all_gather`, `a2a` bshd/sbhd | 4/4 pass — were passing before, and still are |
| CP with FA4 off (control) | 3/3 pass |
| `test_fa4_window_sentinel.py` | 12 passed |
| `test_attention.py` sweep (regression) | 487 passed, 138 skipped, 0 failed |

Numerics are graded against a **float64** reference, never against another backend. That matters
here: the defect survived because the CP suite compares against a run of the same backend, so an
error present on both sides cancels. float64 rather than float32 because torch computes fp32
matmuls in TF32 on Ampere and newer, whose significand is 11 bits — the same as fp16 — so an fp32
reference cannot judge a bf16 kernel.

## Tests

`tests/pytorch/attention/test_fa4_window_sentinel.py`, registered in `qa/L0_pytorch_unittest`:

- eleven parametrized cases for the normalizer, which run on any machine, including that a genuine
  sliding window such as `(511, 0)` is passed through unchanged
- one end-to-end case, skipped unless FA4 is installed on SM100, pinned to FlashAttention and
  anchored to a float64 reference. It asserts a non-zero output separately from the tolerance
  check, so the specific failure is named rather than reported as a large error.

## Note on CI

Three independent reasons this was invisible, each worth addressing on its own and none of them
fixed here:

1. `NVTE_FLASH_ATTN_V4=0` is exported in `qa/L1_pytorch_distributed_unittest/test.sh` and the L0
   suites, so FA4 is never exercised under context parallelism.
2. CI pins `flash-attn-4==4.0.0b11`, which predates Dao-AILab/flash-attention#2490.
3. `run_attention_with_cp.py` grades a CP run against a non-CP run of the same backend, so it
   cannot detect any error affecting both equally. Two of the three comm types reported PASS on a
   backend that was returning zeros.

### CODE REVIEW

File: qa/L0_pytorch_unittest/test.sh [hunk 1]
```diff
@@ -70,6 +70,7 @@ python3 -m pytest --tb=auto --junitxml=$XML_LOG_DIR/pytest_test_fused_mla_q_upro
 python3 -m pytest --tb=auto --junitxml=$XML_LOG_DIR/pytest_test_kv_cache.xml $TE_PATH/tests/pytorch/attention/test_kv_cache.py || test_fail "test_kv_cache.py"
 python3 -m pytest --tb=auto --junitxml=$XML_LOG_DIR/pytest_test_cu_seqlens_cache.xml $TE_PATH/tests/pytorch/attention/test_cu_seqlens_cache.py || test_fail "test_cu_seqlens_cache.py"
 python3 -m pytest --tb=auto --junitxml=$XML_LOG_DIR/pytest_test_mixed_thd_attention.xml $TE_PATH/tests/pytorch/attention/test_mixed_thd_attention.py || test_fail "test_mixed_thd_attention.py"
+python3 -m pytest --tb=auto --junitxml=$XML_LOG_DIR/pytest_test_fa4_window_sentinel.xml $TE_PATH/tests/pytorch/attention/test_fa4_window_sentinel.py || test_fail "test_fa4_window_sentinel.py"
```
[the remark this answers is not in the document]
Comment (nvegesna-netizen) [reply qa/L0_pytorch_unittest/test.sh:73]: Good catch, and the consequence is worse than described — fixed in 927e05ad.

`qa/L0_pytorch_unittest/test.sh:28` does export `NVTE_FLASH_ATTN_V4=0`, but the test would not
merely fail to validate FA4: it would **pass while measuring FA2**. The skip condition checked
`flash_attn_func_v4 is None`, which is an import-time binding that env var does not affect, so the
test would run, TE would decline FA4, and selection would fall through to FA2 — the same lane sets
`NVTE_FLASH_ATTN_V2=1`, and FA2 serves head_dim 128 causal bshd on SM100 perfectly well. A green
test exercising the wrong backend is precisely the failure mode this PR exists to close.

Rather than depend on the lane, the test now pins the backend itself and verifies the result:

```python
pinned = {"NVTE_FUSED_ATTN": "0", "NVTE_UNFUSED_ATTN": "0", "NVTE_FLASH_ATTN_V4": "1"}
...
selected = _attention_backends["flash_attention_backend"]
assert selected is not None and str(selected).startswith("4"), ...
```

The assertion runs before any numerical grading, so the test is correct in any lane and a future
lane cannot quietly opt it out.

I left the registration in L0 rather than moving it to the FA4-enabled L3 matrix, since the test is
now self-sufficient and the eleven non-GPU cases are worth running everywhere. Happy to add it to
L3 as well if you would prefer coverage there too.

File: tests/pytorch/attention/test_fa4_window_sentinel.py
Comment (sudhakarsingh27) [inline tests/pytorch/attention/test_fa4_window_sentinel.py:1]: IMO, it's unnecessary to add the test here, could we add a versioned gate on the calls to "normalize windows" method? (something [like this](https://github.com/NVIDIA/TransformerEngine/pull/3540/changes#diff-001f32a8be891802f01328c1bcc97e377bcc0177f019a1adbbb2ddef627d2520R246-R250) but I'm sure that I've got the minimum version incorrect)

Comment (nvegesna-netizen) [reply tests/pytorch/attention/test_fa4_window_sentinel.py:1]: I checked whether a gate is needed, and I do not think it is.

I transcribed `_resolve_causal_local_window` from before #2490 (`interface.py` @ `59cf537812`) and after it, verified both transcriptions against the real source, and ran every `(causal, window)` pair TE emits through each. TE derives `causal` from `attn_mask_type` and the window from `check_set_window_size`, so the two always agree and `causal=False` with `(-1, 0)` never arises.

| mask type | pre as-is | pre normalized | post as-is | post normalized |
|---|---|---|---|---|
| causal | CAUSAL | CAUSAL | **LOCAL left=-1 right=0** | CAUSAL |
| causal_bottom_right | CAUSAL | CAUSAL | **LOCAL left=-1 right=0** | CAUSAL |
| no_mask | FULL | FULL | FULL | FULL |
| sliding window | LOCAL 511/0 | LOCAL 511/0 | LOCAL 511/0 | LOCAL 511/0 |

Normalizing is inert on the older release, not merely harmless. The old code reaches the same resolved state from `(-1, 0)` by widening that the new code reaches from `(None, 0)` directly, via the same `window_size_left is None and window_size_right == 0` branch, which is unchanged context in #2490's diff. All four entry points funnel through that one resolver on both releases.

So ungated looks safer rather than just simpler: a threshold has to be right, and any release between #2490 merging on 9 Sep and the gated version would stay broken. Reasoning is now in the docstring in `4a84a21e`.

On the test, I would keep it. Eleven of the twelve cases are pure Python with no GPU, so the CI cost is small. What earns their place is that this bug was invisible three separate ways: `NVTE_FLASH_ATTN_V4=0` in the distributed suite, a `flash-attn-4` pin predating #2490, and the CP suite comparing a CP run against a non-CP run of the same backend, which agrees when both sides return zeros. The end to end case pins `NVTE_FLASH_ATTN_V4=1` and asserts the selected backend reports version 4 before grading numbers, so it cannot pass on FA2 the way the original would have.
